relentless/optimize/method.py

```diff
@@ -0,0 +1,193 @@
+"""
+Algorithms
+==========
+
+An :class:`Optimizer` defines an optimization algorithm that can be applied to
+a defined :class:`ObjectiveFunction`.
+
+The following algorithms have been implemented:
+
+.. autosummary::
+    :nosignatures:
+
+    SteepestDescent
+
+.. rubric:: Developer notes
+
+To implement your own optimization algorithm, create a class that derives from
+:class:`Optimizer` and define the required properties and methods.
+
+.. autosummary::
+    :nosignatures:
+
+    Optimizer
+    SteepestDescent
+
+.. autoclass:: Optimizer
+    :member-order: bysource
+    :members: optimize,
+        has_converged,
+        abs_tol
+
+.. autoclass:: SteepestDescent
+    :member-order: bysource
+    :members: optimize,
+        step_size,
+        max_iter
+
+"""
+import abc
+
+from .objective import ObjectiveFunction
+
+class Optimizer(abc.ABC):
+    """Abstract base class for optimization algorithm.
+
+    A :py:class:`Optimizer` defines the optimization algorithm with specified parameters.
+
+    Parameters
+    ----------
+    abs_tol : float or dict
+        The absolute tolerance or tolerances (keyed on the :py:class:`ObjectiveFunction`
+        design variables).
+
+    """
+    def __init__(self, abs_tol):
+        self.abs_tol = abs_tol
+
+    @abc.abstractmethod
+    def optimize(self, objective):
+        """Adjusts the given objective function until the convergence criteria
+        (defined by :py:meth:`has_converged()`) is satisfied.
+
+        Parameters
+        ----------
+        objective : :py:class:`ObjectiveFunction`
+            The objective function to be optimized.
+
+        """
+        pass
+
+    def has_converged(self, result):
+        """Checks if the convergence criteria is satisfied.
```

Comment:
Again, here:
```suggestion
        """Check if the convergence criteria is satisfied.
```
and so on for the rest.